### ISSUE TITLE
Service hairpinning does not work with cilium envoy

### ISSUE THREAD

**Issue by rauanmayemir** [cilium/cilium#28254]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I am trying hairpinning my public domains with coredns to resolve to LB service IP, but it doesn't work. I'm hitting 403 Forbidden error with a message `Access denied` coming from cilium-envoy. Envoy logs have this:
```
[2023-09-23 11:31:39.057][70][warning][filter] [cilium/network_filter.cc:160] [C77533] cilium.network: Policy NOT FOUND for id: 154577 port: 8390
```

I've found prior work on #24536 and #24826 and I'm guessing that gateway api is not assuming the ingress identity.

### Cilium Version

1.14.2

### Kernel Version

5.15.122-flatcar

### Kubernetes Version

1.27.5

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by lmb** [comment]
Hello, thanks for opening the GH issue, please provide steps to reproduce the issue and a cilium sysdump otherwise the issue can't be fixed. Thank you.

**Comment by rauanmayemir** [comment]
@lmb Sent you a sysdump link privately on cilium's slack.

I've managed to come up with a simple repro for the [getting started demo](https://docs.cilium.io/en/stable/gettingstarted/demo/):

```
kubectl create -f https://raw.githubusercontent.com/cilium/cilium/1.14.2/examples/minikube/http-sw-app.yaml
```

Let's set up our routes. (assuming cilium runs with envoy in DS mode and gateway api is enabled)

```
---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: cilium-hairpin
  namespace: default
spec:
  gatewayClassName: cilium
  listeners:
    - protocol: HTTP
      port: 80
      name: http
      hostname: "hairpinning.test"
      allowedRoutes:
        namespaces:
          from: All
---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: hairpinning
  namespace: default
spec:
  parentRefs:
    - name: cilium-hairpin
  hostnames:
    - hairpinning.test
  rules:
    - backendRefs:
        - name: deathstar
          port: 80
```

Now we try calling our endpoint replacing $LB_IP with our gateway's LB external IP:

```shell
kubectl exec xwing -- curl -vv -k -X POST http://hairpinning.test/v1/request-landing --resolve hairpinning.test:80:$LB_IP
```

This is what I'm getting back on my end:
```
kubectl exec xwing -- curl -v -X POST http://hairpinning.test/v1/request-landing --resolve hairpinning.test:80:192.168.200.116
* Added hairpinning.test:80:192.168.200.116 to DNS cache
* Hostname hairpinning.test was found in DNS cache
*   Trying 192.168.200.116...
* TCP_NODELAY set
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to hairpinning.test (192.168.200.116) port 80 (#0)
> POST /v1/request-landing HTTP/1.1
> Host: hairpinning.test
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 403 Forbidden
< content-length: 15
< content-type: text/plain
< date: Tue, 26 Sep 2023 11:40:59 GMT
< server: envoy
<
{ [15 bytes data]
* Curl_http_done: called premature == 0
100    15  100    15    0     0   6768      0 --:--:-- --:--:-- --:--:-- 15000
* Connection #0 to host hairpinning.test left intact
Access denied
```

Initially I thought it was only an issue for routes with TLS, but it actually failed same for regular routes.
I didn't know about this because I have a default route with RequestRedirect filter from insecure to https and it didn't actually fail with 403 but correctly returned '301 Moved permanently'. But when I served the actual content instead, envoy failed with 403.

Here's what I had for HTTPS:
```
curl -vv -k -X POST https://hairpinning.test/v1/request-landing --resolve hairpinning.test:443:192.168.200.216
* Added hairpinning.test:443:192.168.200.216 to DNS cache
* Hostname hairpinning.test was found in DNS cache
*   Trying 192.168.200.216...
* TCP_NODELAY set
* Connected to hairpinning.test (192.168.200.216) port 443 (#0)
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=hairpinning.test
*  start date: Sep 26 11:04:32 2023 GMT
*  expire date: Dec 25 11:04:32 2023 GMT
*  issuer: CN=hairpinning.test
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
> POST /v1/request-landing HTTP/1.1
> Host: hairpinning.test
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 403 Forbidden
< content-length: 15
< content-type: text/plain
< date: Tue, 26 Sep 2023 11:21:05 GMT
< server: envoy
< 
Access denied
* Curl_http_done: called premature == 0
* Connection #0 to host hairpinning.test left intact
```

With cilium envoy logs showing:

```
[2023-09-26 11:28:31.687][119][warning][filter] [cilium/network_filter.cc:160] [C712] cilium.network: Policy NOT FOUND for id: 134610 port: 80
```

134610 - is the cilium identity of deathstar.

**Comment by rauanmayemir** [comment]
I could not find a way to properly anonymize the sysdump logs to share it publicly.
I believe I provided all the information for reproducing the issue, if it's not enough, I will regretfully give up.

**Comment by rauanmayemir** [comment]
Not stale. Still reproducible on `1.14.3`.

**Comment by rauanmayemir** [comment]
Not stale.

**Comment by youngnick** [comment]
Okay, @rauanmayemir, I had a chance to try and reproduce this in kind, and I could not.

I set up a kind cluster, installed the application and added the Gateway config you have above.

The Gateway works reliably from inside and outside the cluster.

I don't have any network policy though, and based your comments on #30073, it seems likely you do. Could you please supply any network policy you have in the cluster?

**Comment by rauanmayemir** [comment]
@youngnick I have so many policies, I don’t know where to start. 🥲 

I think I will go and try to reproduce it on vanilla kind cluster as well.

I wonder if it could have anything to do with host firewalls and do I need to expicitly allow ingress ports. (it just hit me recently, but it should have been registered as an audit, not outright drop)

**Comment by youngnick** [comment]
Yes, I've been testing for the other issue and have seen drops _from_ the ingress identity _to_ the external LB IP (since it's `world` identity). I'm just working up some recommendations now. But if you have a default deny somewhere, probably you're having something similar.

**Comment by youngnick** [comment]
Okay, I've found a ClusterwideNetworkPolicy that fixes the issue on my kind cluster. @rauanmayemir, could you try this out?

```yaml
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: allow-ingress-to-cluster
  namespace: default
spec:
  endpointSelector:
    matchExpressions:
    - key: reserved:ingress
      operator: Exists
  egress:
  - toEntities:
    - cluster
```

This allows traffic _from_ the reserved `ingress` identity that was introduced recently _to_ any entity in the cluster. I _think_ this is what is causing this issue.

If this does fix your issue, then I think a documentation PR will be in order.

**Comment by rauanmayemir** [comment]
I'm on `1.14.3`, I don't believe it can fix it for me because `ingress` identity was merged later. I tried it anyway, and it doesn't fix my issue.

But I think this is indeed the issue with my LB service having a VIP outside of kubernetes cidr.

`curl -vv 'https://domain.com' --resolve 'domain:443:192.168.200.41'`  query above results in the following log line in `cilium-envoy`:

```
[2024-03-12 05:20:20.719][84][warning][filter] [cilium/network_filter.cc:160] [C3740627] cilium.network: Policy NOT FOUND for id: 2 port: 30080
```

id: 2 is `reserved:world` and port 30080 is indeed a pod port for the workload serving 'domain.com'. It might help if I upgrade to 1.15.x and set up policies for ingress.

**Comment by rauanmayemir** [comment]
@youngnick Upgraded to `v1.15.4` and it still doesn't work, failing with 403. (but this time there are no warnings in envoy)
Adding egress to cluster from `reserved:ingress` unblocked global ingress, but did not help with intra-cluster hairpinning.

**Comment by sayboras** [comment]
Thanks for your comment. 

On the hair-pinning issue, can you share cilium config map as it might be fixed as part of #32367? Thanks.

**Comment by rauanmayemir** [comment]
@sayboras Here's my cilium cm:

<details>
  <summary>cilium-configmap</summary>

```
# Identity allocation mode selects how identities are shared between cilium
# nodes by setting how they are stored. The options are "crd" or "kvstore".
# - "crd" stores identities in kubernetes as CRDs (custom resource definition).
#   These can be queried with:
#     kubectl get ciliumid
# - "kvstore" stores identities in an etcd kvstore, that is
#   configured below. Cilium versions before 1.6 supported only the kvstore
#   backend. Upgrades from these older cilium versions should continue using
#   the kvstore by commenting out the identity-allocation-mode below, or
#   setting it to "kvstore".
identity-allocation-mode: crd
identity-heartbeat-timeout: "30m0s"
identity-gc-interval: "15m0s"
cilium-endpoint-gc-interval: "5m0s"
nodes-gc-interval: "5m0s"
skip-cnp-status-startup-clean: "false"

# If you want to run cilium in debug mode change this value to true
debug: "false"
debug-verbose: ""
# The agent can be put into the following three policy enforcement modes
# default, always and never.
# https://docs.cilium.io/en/latest/security/policy/intro/#policy-enforcement-modes
enable-policy: "always"
policy-cidr-match-mode: ""
# If you want metrics enabled in all of your Cilium agents, set the port for
# which the Cilium agents will have their metrics exposed.
# This option deprecates the "prometheus-serve-addr" in the
# "cilium-metrics-config" ConfigMap
# NOTE that this will open the port on ALL nodes where Cilium pods are
# scheduled.
prometheus-serve-addr: ":9962"
# A space-separated list of controller groups for which to enable metrics.
# The special values of "all" and "none" are supported.
controller-group-metrics:
  write-cni-file
  sync-host-ips
  sync-lb-maps-with-k8s-services
# If you want metrics enabled in cilium-operator, set the port for
# which the Cilium Operator will have their metrics exposed.
# NOTE that this will open the port on the nodes where Cilium operator pod
# is scheduled.
operator-prometheus-serve-addr: ":9963"
enable-metrics: "true"
enable-envoy-config: "true"
enable-gateway-api: "true"
enable-gateway-api-secrets-sync: "true"
gateway-api-secrets-namespace: "cilium-secrets"
loadbalancer-l7: "envoy"
loadbalancer-l7-ports: ""
loadbalancer-l7-algorithm: "round_robin"

# Enable IPv4 addressing. If enabled, all endpoints are allocated an IPv4
# address.
enable-ipv4: "true"

# Enable IPv6 addressing. If enabled, all endpoints are allocated an IPv6
# address.
enable-ipv6: "false"
# Users who wish to specify their own custom CNI configuration file must set
# custom-cni-conf to "true", otherwise Cilium may overwrite the configuration.
custom-cni-conf: "false"
enable-bpf-clock-probe: "false"
# If you want cilium monitor to aggregate tracing for packets, set this level
# to "low", "medium", or "maximum". The higher the level, the less packets
# that will be seen in monitor output.
monitor-aggregation: medium

# The monitor aggregation interval governs the typical time between monitor
# notification events for each allowed connection.
#
# Only effective when monitor aggregation is set to "medium" or higher.
monitor-aggregation-interval: "5s"

# The monitor aggregation flags determine which TCP flags which, upon the
# first observation, cause monitor notifications to be generated.
#
# Only effective when monitor aggregation is set to "medium" or higher.
monitor-aggregation-flags: all
# Specifies the ratio (0.0-1.0] of total system memory to use for dynamic
# sizing of the TCP CT, non-TCP CT, NAT and policy BPF maps.
bpf-map-dynamic-size-ratio: "0.0025"
# bpf-policy-map-max specifies the maximum number of entries in endpoint
# policy map (per endpoint)
bpf-policy-map-max: "16384"
# bpf-lb-map-max specifies the maximum number of entries in bpf lb service,
# backend and affinity maps.
bpf-lb-map-max: "65536"
bpf-lb-external-clusterip: "true"

# Pre-allocation of map entries allows per-packet latency to be reduced, at
# the expense of up-front memory allocation for the entries in the maps. The
# default value below will minimize memory usage in the default installation;
# users who are sensitive to latency may consider setting this to "true".
#
# This option was introduced in Cilium 1.4. Cilium 1.3 and earlier ignore
# this option and behave as though it is set to "true".
#
# If this value is modified, then during the next Cilium startup the restore
# of existing endpoints and tracking of ongoing connections may be disrupted.
# As a result, reply packets may be dropped and the load-balancing decisions
# for established connections may change.
#
# If this option is set to "false" during an upgrade from 1.3 or earlier to
# 1.4 or later, then it may cause one-time disruptions during the upgrade.
preallocate-bpf-maps: "false"

# Regular expression matching compatible Istio sidecar istio-proxy
# container image names
sidecar-istio-proxy-image: "cilium/istio_proxy"

# Name of the cluster. Only relevant when building a mesh of clusters.
cluster-name: my-cluster
# Unique ID of the cluster. Must be unique across all conneted clusters and
# in the range of 1 and 255. Only relevant when building a mesh of clusters.
cluster-id: "2"

# Encapsulation mode for communication between nodes
# Possible values:
#   - disabled
#   - vxlan (default)
#   - geneve
routing-mode: "tunnel"
tunnel-protocol: "geneve"
service-no-backend-response: "reject"


# Enables L7 proxy for L7 policy enforcement and visibility
enable-l7-proxy: "true"

enable-ipv4-masquerade: "true"
enable-ipv4-big-tcp: "false"
enable-ipv6-big-tcp: "false"
enable-ipv6-masquerade: "true"
enable-bpf-masquerade: "true"
enable-masquerade-to-route-source: "false"

enable-xt-socket-fallback: "true"
install-no-conntrack-iptables-rules: "false"

auto-direct-node-routes: "false"
enable-bandwidth-manager: "true"
enable-bbr: "false"
enable-local-redirect-policy: "true"
enable-host-firewall: "true"

kube-proxy-replacement: "true"
kube-proxy-replacement-healthz-bind-address: ""
bpf-lb-sock: "false"
bpf-lb-sock-hostns-only: "true"
enable-health-check-nodeport: "true"
enable-health-check-loadbalancer-ip: "false"
node-port-bind-protection: "true"
enable-auto-protect-node-port-range: "true"
bpf-lb-acceleration: "disabled"
enable-service-topology: "true"
enable-svc-source-range-check: "true"
enable-l2-neigh-discovery: "true"
arping-refresh-period: "30s"
log-system-load: "true"
k8s-require-ipv4-pod-cidr: "true"
enable-k8s-networkpolicy: "true"
# Tell the agent to generate and write a CNI configuration file
write-cni-conf-when-ready: /host/etc/cni/net.d/05-cilium.conflist
cni-exclusive: "true"
cni-log-file: "/var/run/cilium/cilium-cni.log"
enable-endpoint-health-checking: "true"
enable-health-checking: "true"
enable-well-known-identities: "true"
enable-remote-node-identity: "true"
synchronize-k8s-nodes: "true"
policy-audit-mode: "true"
operator-api-serve-addr: "127.0.0.1:9234"
# Enable Hubble gRPC service.
enable-hubble: "true"
# UNIX domain socket for Hubble server to listen to.
hubble-socket-path: "/var/run/cilium/hubble.sock"
# Address to expose Hubble metrics (e.g. ":7070"). Metrics server will be disabled if this
# field is not set.
hubble-metrics-server: ":9965"
# A space separated list of metrics to enable. See [0] for available metrics.
#
# https://github.com/cilium/hubble/blob/master/Documentation/metrics.md
hubble-metrics:
  dns:query;ignoreAAAA
  drop
  tcp
  flow
  flows-to-world:any-drop;
  port-distribution
  icmp
  httpV2:exemplars=true;labelsContext=source_ip,source_namespace,source_workload,destination_ip,destination_namespace,destination_workload,traffic_direction
enable-hubble-open-metrics: "true"
hubble-export-file-max-size-mb: "10"
hubble-export-file-max-backups: "5"
# An additional address for Hubble server to listen to (e.g. ":4244").
hubble-listen-address: ":4244"
hubble-disable-tls: "false"
hubble-tls-cert-file: /var/lib/cilium/tls/hubble/server.crt
hubble-tls-key-file: /var/lib/cilium/tls/hubble/server.key
hubble-tls-client-ca-files: /var/lib/cilium/tls/hubble/client-ca.crt
ipam: "cluster-pool"
ipam-cilium-node-update-rate: "15s"
cluster-pool-ipv4-cidr: "172.33.0.0/16"
cluster-pool-ipv4-mask-size: "22"
egress-gateway-reconciliation-trigger-interval: "1s"
enable-vtep: "false"
vtep-endpoint: ""
vtep-cidr: ""
vtep-mask: ""
vtep-mac: ""
enable-bgp-control-plane: "true"
bgp-secrets-namespace: "kube-system"
enable-pmtu-discovery: "true"
procfs: "/host/proc"
bpf-root: "/sys/fs/bpf"
cgroup-root: "/run/cilium/cgroupv2"
enable-k8s-terminating-endpoint: "true"
enable-sctp: "false"
annotate-k8s-node: "true"

k8s-client-qps: "5"
k8s-client-burst: "10"
remove-cilium-node-taints: "true"
set-cilium-node-taints: "true"
set-cilium-is-up-condition: "true"
unmanaged-pod-watcher-interval: "15"
# default DNS proxy to transparent mode in non-chaining modes
dnsproxy-enable-transparent-mode: "true"
tofqdns-dns-reject-response-code: "refused"
tofqdns-enable-dns-compression: "true"
tofqdns-endpoint-max-ip-per-hostname: "50"
tofqdns-idle-connection-grace-period: "0s"
tofqdns-max-deferred-connection-deletes: "10000"
tofqdns-proxy-response-max-delay: "100ms"
agent-not-ready-taint-key: "node.cilium.io/agent-not-ready"

mesh-auth-enabled: "false"
mesh-auth-queue-size: "1024"
mesh-auth-rotated-identities-queue-size: "1024"
mesh-auth-gc-interval: "5m0s"

proxy-connect-timeout: "2"
proxy-max-requests-per-connection: "0"
proxy-max-connection-duration-seconds: "0"

external-envoy-proxy: "true"
max-connected-clusters: "255"

# Extra config allows adding arbitrary properties to the cilium config.
# By putting it at the end of the ConfigMap, it's also possible to override existing properties.

allow-localhost: policy
```
</details>

**Comment by rauanmayemir** [comment]
@sayboras @youngnick I'm sorry for the confusion, I went and tried something else today:

```
  - endpointSelector:
      matchExpressions:
        - key: reserved:ingress
          operator: Exists
    egress:
      - toEntities:
          - cluster
    ingress:
      - fromEntities:
          - world
          - cluster
```

This works! 💃
I was going to review host firewall monitor verdicts from the agent today, but fortunately found the bottleneck. I think this issue is solved. 😌

Not only do we need egress rule into `cluster`, but also ingress _from_ `cluster` (seems so obvious in hindsight). This 'trick' will mean a lot for gateway-api GAMMA initiative.

**Comment by sayboras** [comment]
> Here's my cilium cm

Can you give it a try with the image from below PR? Seems like you are having bpf.masquerade enabled, one additional config `bpf.hostLegacyRouting=true` could be required as a workaround.

https://github.com/cilium/cilium/pull/32367
